src/NServiceBus.Core/Utils/MsmqExtensions.cs

```diff
@@ -0,0 +1,174 @@
+﻿namespace NServiceBus.Utils
+{
+    using System;
+    using System.ComponentModel;
+    using System.Messaging;
+    using System.Runtime.InteropServices;
+    using System.Security.Principal;
+
+    static class MsmqExtensions
+    {
+        const string Mqrt = "mqrt.dll";
+        const string Advapi32 = "advapi32.dll";
+
+        [DllImport(Mqrt, CharSet = CharSet.Unicode, SetLastError = true)]
+        static extern int MQGetQueueSecurity(string formatName, int SecurityInformation, IntPtr SecurityDescriptor, int length, out int lengthNeeded);
+
+        [DllImport(Advapi32, SetLastError = true)]
+        static extern bool GetSecurityDescriptorDacl(IntPtr pSD, out bool daclPresent, out IntPtr pDacl, out bool daclDefaulted);
+
+        [DllImport(Advapi32, CharSet = CharSet.Unicode, SetLastError = true)]
+        static extern bool GetAclInformation(IntPtr pAcl, ref ACL_SIZE_INFORMATION pAclInformation, uint nAclInformationLength, ACL_INFORMATION_CLASS dwAclInformationClass);
+
+        [DllImport(Advapi32, CharSet = CharSet.Unicode, SetLastError = true)]
+        static extern int GetAce(IntPtr aclPtr, int aceIndex, out IntPtr acePtr);
+
+        [DllImport(Advapi32, CharSet = CharSet.Unicode, SetLastError = true)]
+        static extern int GetLengthSid(IntPtr pSID);
+
+        [DllImport(Advapi32, CharSet = CharSet.Unicode, SetLastError = true)]
+        static extern bool ConvertSidToStringSid([MarshalAs(UnmanagedType.LPArray)] byte[] pSID, out IntPtr ptrSid);
+
+        const int DACL_SECURITY_INFORMATION = 4;
+
+        //Security constants
+        const int MQ_ERROR_SECURITY_DESCRIPTOR_TOO_SMALL = unchecked((int) 0xc00e0023);
+        const int MQ_OK = 0;
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct ACE_HEADER
+        {
+            public byte AceType;
+            public byte AceFlags;
+            public short AceSize;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct ACCESS_ALLOWED_ACE
+        {
+            public ACE_HEADER Header;
+            public uint Mask;
+            public int SidStart;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        struct ACL_SIZE_INFORMATION
+        {
+            public uint AceCount;
+            public uint AclBytesInUse;
+            public uint AclBytesFree;
+        }
+
+        enum ACL_INFORMATION_CLASS
+        {
+            AclRevisionInformation = 1,
+            AclSizeInformation
+        }
+        
+        public static bool TryGetPermissions(this MessageQueue queue, string user, out MessageQueueAccessRights? rights)
+        {
+            string sid = GetSidForUser(user);
+
+            try
+            {
+                rights = GetPermissions(queue.FormatName, sid);
+                return true;
+            }
+            catch
```

Comment:
@SimonCropp Possibly, I need to go over this code again to comment it a bit (most of it taken from referencesource of `MessageQueue` class, some from StackOverflow). There were some exceptions thrown that were internal in the framework, so I threw a plain old Exception instead. I'll fix it.